### PR TITLE
Couchbase service & ticket registries

### PR DESCRIPTION
**Pull request by mmoayyed**
Closes https://github.com/Jasig/cas/issues/1059

### CODE REVIEW

File: cas-server-support-couchbase-core/src/main/java/org/jasig/cas/couchbase/core/CouchbaseClientFactory.java [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    /**
+     * Instantiates a new Couchbase client factory.
+     */
+    public CouchbaseClientFactory() {}
+
+    /**
+     * Start initializing the client. This will schedule a task that retries
+     * connection until successful.
+     */
+    public void initialize() {
```
Comment (leleuj) [inline cas-server-support-couchbase-core/src/main/java/org/jasig/cas/couchbase/core/CouchbaseClientFactory.java]: How is this method called?

Comment (mmoayyed) [inline cas-server-support-couchbase-core/src/main/java/org/jasig/cas/couchbase/core/CouchbaseClientFactory.java]: This is called by either the service/ticket registry components @PostConstruct methods.

**Comment by mmoayyed** [comment]
I am still struggling to find how to write unit tests for this. There is a mock server, but it is way too buggy.

**Comment by mmoayyed** [comment]
https://github.com/couchbase/CouchbaseMock/issues/16 and https://github.com/couchbase/CouchbaseMock/issues/11

**Comment by mmoayyed** [comment]
and apparently, it is impossible to find a cloud solution that offers basic couchdb hosting.

**Comment by leleuj** [comment]
You're right: it doesn't seem there is a Couchbase cloud provider. Very strange and disturbing. I guess you can still setup a micro server on AWS based on a community couchbase server image (one year free if you don't have already used it).

**Comment by mmoayyed** [comment]
I may go the AWS route, sure. I might also ask @fjollberg to test the patch here for us. 

Also note that there are a few differences between this PR and the original work. I have intentionally removed them until I can confirm the basic functionality with a test instance, and will gradually add those back.

**Comment by fjollberg** [comment]
Before I converted the code to the new CAS service JSON serialiser I had some unit tests verifying the serialiser. The layer on top of couchbase is thin enough to make unit testing almost pointless, there is hardly any logic there at all. The stuff that's interesting to test is the interaction with the couchbase library and server, so eventually I just dropped the idea of unit testing.

**Comment by fjollberg** [comment]
I guess the bits that are removed are the asynchronous attachment to the couchbase server. It can really be argued if you want that in there at all since it's a somewhat different behaviour than you may otherwise expect. It allows the CAS server to start even if the couchbase server hasn't quite started yet, and attach to the couchbase server once it is running.

In our implementation we have a separate monitoring servlet in our overlay which detects and indicates this situation which our load balancer is polling.

A different approach and perhaps more straight forward is to assume that couchbase is up and running and instead test for that before the CAS container startup. But the way we did it worked well enough for us to leave it as it is and made server boot procedure a bit faster and simpler.

**Comment by mmoayyed** [comment]
What I can do for unit tests is to at least make them available, so that once we need to troubleshoot something, at least we have the tests for it, and we could then just connect the code either to AWS or a local install. Similar to yours, I mostly just want to ensure the serializer works correctly, should we decide to augment the JSON model in the future. 

As for starting the cluster in an async mode, that is a valid point. I just don't know if it really makes sense to start a CAS server without a functional service registry. That seems odd to me, so I might take you up on your suggestion to check the connection upfront and fail to startup altogether so the deployer is immediately made aware. At the same time, we want to make sure CAS is still functional when the DB is taken down (similar to how JPA works) so I'll make sure to account for that. 

The other bit that I have removed for now is the ability to configure a static list of services upfront. That didnt make much sense to me. Why would you need that? 

Thanks again. I am sure this will prove very useful to the CAS community.

**Comment by fjollberg** [comment]
I think you don't want the async attach really. I've not investigated this aspect of the most recent couchbase client. In the earlier version it was just on initial startup it would fail if unable to find any living server. If running it would reattach nicely.

**Comment by fjollberg** [comment]
The static list was also a result of our environment. We've used it to hard wire the management service itself into the registry in order to prevent configuration mistakes locking it out, at the same time preventing the service from starting with an empty service list, i.e. in open mode.

**Comment by mmoayyed** [comment]
Ah, OK. That bit I think we can safely drop. The management interface on startup auto-adds itself to the registry, and CAS does no longer permit open mode for the registry. So we should be good on that front. Does everything else in the pull look ok to you?

**Comment by mmoayyed** [comment]
> it would fail if unable to find any living server. If running it would reattach nicely.

Perfect. Thanks.

I have added additional unit tests and have gotten access to a couchbase install finally, which was super easy to do. I'll get this merged, and then will run a few tests to ensure we are good here. Thanks again!

**Pull request merged**